Defense channels and rolemapping

zuuring opened the pull request:
### Notes
This adds a new discord command `defense-audit` which has one parameter `client-name` which will then perform the following steps:

- Create two channels, `int-clientname-audit` and `ext-clientname-audit` 
- Create two new roles `Defense Internal: clientname` and `Defense Extenal: clientname`
- Grant `ViewChannel` access to both channels for `Defense Internal: *` and only `ext-*-audit` channel access to `Defense Extenal`
- Generate an invite link for both channels

### Screenshot
<img width="702" alt="Screenshot 2024-03-20 at 13 51 15" src="https://github.com/thesis/valkyrie/assets/7145746/7abb64fd-3476-47db-96ec-55d2f8a88f42">


### Automatic role assignment
Also included in this PR is automatic role assignment based of the invite code used. Due to lack of IDs associated to `GuildMemberAdd` events when a new user joins discord, the only way to tie a discord user to an invite link is to count invite code uses within the guild and compare it once a new member has joined the discord. For this to work, we check the invites and match role assignment to channel.name the invite link was generated from. 

### To-do
- [x] Setup `defense-audit` command
- [x] Add rolemapping based on invite joins
- [x] Fix issue on `int-client-name-audit` channels not finding correct role for `ext-client-name-audit` channels

zuuring commented on discord-scripts/invite-management.ts:
```diff
@@ -133,10 +209,86 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
+            ? channelTypeMatch[2]
+                .replace(/-/g, " ")
+                .split(" ")
+                .map(
```
Going to refactor this around to just match ChannelID -> ChannelName -> Assigned roles and then grab role with `Defense External: Client` rather than falling overmyself with these consts 😄

zuuring commented on discord-scripts/invite-management.ts:
```diff
@@ -133,10 +209,86 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
+      )
+      guildInvites.set(member.guild.id, newInvites)
+
+      robot.logger.info(
```
Will streamline this to reduce loggin for debugging and fix up `oldInvites` and `newInvites`

Shadowfiend commented:
Question here: can we kill the embeds in the intro message?

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -1,8 +1,15 @@
 import { DAY, MILLISECOND, WEEK } from "../lib/globals.ts"
 
-const EXTERNAL_AUDIT_CHANNEL_REGEXP = /^ext-(?<client>.*)-audit$/
+const guildInvites = new Collection()
```
Any reason not to just use objects here?

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -22,10 +29,34 @@ async function createInvite(
 
+async function listInvites(discordClient: Client, robot: Robot): Promise<void> {
+  discordClient.guilds.cache.forEach(async (guild) => {
+    const fetchInvites = await guild.invites.fetch().catch((error) => {
+      robot.logger.error(
+        `Failed to fetch invites for guild ${guild.name}: ${error}`,
+      )
+    })
```
When we use `async`/`await`, let's use `try`/`catch` rather than `.catch` behavior. It'll clean up the control flow in this whole function.

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -38,6 +69,28 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
+        description: "Creates Defense audit channels",
+        options: [
+          {
+            name: "client-name",
```
Let's call this `audit-name`.

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -38,6 +69,28 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
+          {
+            name: "client-name",
+            type: ApplicationCommandOptionType.String,
+            description: "The name of the client to create the channels for",
```
Let's say “The name of the audit/client to create the channels for.”

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -75,67 +128,183 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
-              reason: `Role for ${channel.name} channel`,
-            })
+      if (
+        !interaction.guild ||
```
Can't happen after the previous statement, right?

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -75,67 +128,183 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
+      ) {
+        await interaction.reply({
+          content:
+            "This command can only be run in channels under the 'defense' category.",
```
```suggestion
            "This command can only be run in chat channels under the 'Defense' category.",
```

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -75,67 +128,183 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
 
-            await channel.permissionOverwrites.create(role, {
-              ViewChannel: true,
+      const clientName = interaction.options.get("client-name")
+      if (!clientName) {
+        await interaction.reply({
+          content: "Client name is required for the defense-audit command.",
+          ephemeral: true,
+        })
+        return
+      }
```
To be updated based on the above comment as well.

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -75,67 +128,183 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
+        if (typeof clientName.value === "string") {
+          await interaction.deferReply({ ephemeral: true })
+          const normalizedClientName = clientName.value
+            .replace(/\s+/g, "-")
```
Do we have any other limitations on names here? e.g. special characters. Wonder if this is more like `.replace(/[^a-zA-Z0-9]/g, "-")`.

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -75,67 +128,183 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
+          const externalChannelName = `ext-${normalizedClientName}-audit`
+
+          const defenseCategory = interaction.guild.channels.cache.find(
+            (c) => c.name === "defense",
```
`c.type` should also be a category, right? I don't think we have a plain `#defense` channel, but I can imagine a scenario where we would.

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -75,67 +128,183 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
+            return
+          }
+
+          const internalChannel = await interaction.guild.channels.create({
+            name: internalChannelName,
+            type: ChannelType.GuildText,
+            parent: defenseCategory.id,
+          })
+
+          const externalChannel = await interaction.guild.channels.create({
+            name: externalChannelName,
+            type: ChannelType.GuildText,
+            parent: defenseCategory.id,
+          })
```
Let's check if these exist before creating—i.e. do more of a `createOrUpdate` operation where we create it if it exists, and if it doesn't then we sync permissions.

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -75,67 +128,183 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
+            parent: defenseCategory.id,
+          })
+
+          const internalRoleName = `Defense Internal: ${clientName.value}`
+          const externalRoleName = `Defense External: ${clientName.value}`
+          const internalRole = await interaction.guild.roles.create({
+            name: internalRoleName,
+            reason: "Role for internal audit channel",
+          })
+          const externalRole = await interaction.guild.roles.create({
+            name: externalRoleName,
+            reason: "Role for external audit channel",
+          })
```
Same deal here—though maybe `.create` on the Discord API already does this?

zuuring replied:
Switched this one around as it seems like actually discord `.create` does not, it will make an ever infinite amount of roles with the same name! Also the case for channel names, I am seeing 😆

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -75,67 +128,183 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
+          })
+
+          const internalInvite = await createInvite(internalChannel)
+          const externalInvite = await createInvite(externalChannel)
```
For clarity, I would do all of the internal channel stuff first, and then all of the external channel stuff. That'll make it easier to tell what we're doing to which channel, and identify if we've messed up in the future.

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -75,67 +128,183 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
+          const internalInvite = await createInvite(internalChannel)
+          const externalInvite = await createInvite(externalChannel)
+
+          await interaction.editReply({
+            content:
+              `**Defense audit setup complete for: ${clientName.value}**\n\n` +
+              `Internal Channel: <#${internalChannel.id}> (Invite: ${internalInvite.url})\n` +
+              `External Channel: <#${externalChannel.id}> (Invite: ${externalInvite.url})\n\n` +
+              `Roles created: <@&${internalRole.id}>, <@&${externalRole.id}>`,
+          })
+        }
```
We should still send the initial message to the internal and external channels that we were sending before—it'll help to have a reference point inside the channels with the invite/permissions info.

zuuring replied:
Setup a @here message to be sent to the channels being setup, we could also tag the initiated user if needed.

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -75,67 +128,183 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
+                .join(" ")
+            : ""
+
+          robot.logger.info(`Channel Name: ${channel.name}`)
```
If we keep this log we should give it context as to what process is logging it.

zuuring replied:
Added a better logger at the end of this, so we see when invite is used + role matched to.

Shadowfiend reviewed (changes requested):
Left a handful of questions & comments.

zuuring commented:
@Shadowfiend Just pushed an update to this with fixes to the review comments as following: 
- Update permissions if channel already exists
- Check if role already exists, if so skip
- Changed logic so that if audit channels already exist, we output a different message to interaction
- Remove embeds by wrapping invite URL in quotes.
- Change verbiage around messages
- Separates InternalChannel and externalChannel
- Changed invite code permission match

Now one thing that came up, turns out Discord channel categories have a limit of max (50) channels per category. Should we maybe setup an "Audit" category or even (not ideal) a channel category on a per client basis?

Shadowfiend commented:
> Now one thing that came up, turns out Discord channel categories have a limit of max (50) channels per category. Should we maybe setup an "Audit" category or even (not ideal) a channel category on a per client basis?

Lawd. I think we can hold for now, as I don't think we're going to have 20 audits going in parallel, and older audits will get archived. Making them their own category is a good idea though and we can discuss with Defense once we've got the basics working here.

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -31,19 +30,19 @@ async function createInvite(
+    try {
+      const fetchInvites = await guild.invites.fetch()
+      if (fetchInvites) {
+        guildInvites[guild.id] = guildInvites[guild.id] || {}
```
Fancy-shmancy here is `guildInvites[guild.id] ??= {}`.

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -172,13 +172,13 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
             .toLowerCase()
-          const internalChannelName = `int-${normalizedClientName}-audit`
-          const externalChannelName = `ext-${normalizedClientName}-audit`
+          const internalChannelName = `🔒int-${normalizedClientName}-audit`
+          const externalChannelName = `🔒ext-${normalizedClientName}-audit`
```
IMO we should let the existing automation handle this, unless it's not working for some reason.

zuuring replied:
This was a weird one where it seemed to not work like it should of without this definition and 🔒 prefix.

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -189,49 +189,98 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
+          let internalChannel = interaction.guild.channels.cache.find(
+            (channel) => channel.name === internalChannelName,
+          ) as TextChannel
+          const internalChannelCreated = !internalChannel
+          if (internalChannelCreated) {
```
This should be `internalChannelNeedsCreation` or something, right?

Shadowfiend commented on discord-scripts/invite-management.ts:
```diff
@@ -189,49 +189,98 @@ export default async function sendInvite(discordClient: Client, robot: Robot) {
+              ViewChannel: true,
+            })
+            await internalChannel.send(
+              `@here **Welcome to the ${clientName.value} Internal Audit Channel!**`,
```
Sorry, what I meant with this message is that we should keep the full original version. i.e.:

```
**Welcome to the ${clientName.value} audit internal channel!**

You can use this invite to ...: ${internalInvite.url}.
```

And similar for external except it would say “Welcome to the ${clientName.value} audit channel...” (note that it doesn't say “external”, it just doesn't say “internal”).

The goal here is to have a point (possibly pinned) in each channel where folks can:
- See the invite code.
- See information about the invite code (e.g. when it expires and max uses).

Basically the same stuff we were showing @ https://github.com/thesis/valkyrie/pull/301/files#diff-b34d961b5563321a08dc8493fd258c4f5cec44c90e8edd92999c529b43b0f7aaL87-L98 .

zuuring replied:
That makes sense and will get this changed over! Only question I have about that, if we pin the invite information to each channel, should we worry about deleting that message once a client joins or just simply always have it visible, and just assume team members can remove message if needed. I love the expiry / uses idea, that makes a lot of sense and also helps track.

Another layer, we could even update pinned message with a `@user` of who has claimed invite? If it seems important to surface.

Shadowfiend replied:
Love the idea of updating users & counts, though I would say not top priority.

Shadowfiend reviewed (changes requested):
Think I'm going to hold this up on the remaining remarks—it's close though!

zuuring commented:
@Shadowfiend Just pushed an update on this with the fixes based on your review! I also added a way to show expiry date using Discord's built in time rendering. That way we get a nice hover event when going over the expiry time.
<img width="275" alt="Screenshot 2024-04-22 at 16 06 29" src="https://github.com/thesis/valkyrie/assets/7145746/794ae50f-9797-4b2b-9531-e73d91aba7e8">

Shadowfiend reviewed (approved):
Changes look good, let's :shipit:
